test/sketch_test.cpp

```diff
@@ -65,57 +65,125 @@ TEST(SketchTestSuite, GIVENonlyIndexZeroUpdatedTHENitWorks) {
   // THEN it works
   Update res = sketch.query();
   Update exp {0,delta};
-  ASSERT_EQ(res, exp) << "Expected: " << exp << "\nActual: " << res;
+  ASSERT_EQ(res, exp) << "Expected: " << exp << std::endl << "Actual: " << res;
 }
 
-TEST(SketchTestSuite, TestingSketchAddition){
-  srand (time(NULL));
-  for (int i = 0; i < 1000; i++){
-    const unsigned long vect_size = 1000;
-    const unsigned long num_updates = 1000;
-    Sketch sketch = Sketch(vect_size,rand());
-    Testing_Vector test_vector = Testing_Vector(vect_size,num_updates);
+/**
+ * Make sure sketch sampling works
+ */
+void test_sketch_sample(unsigned long num_sketches,
+  unsigned long vec_size, unsigned long num_updates,
+  double max_sample_fail_prob, double max_bucket_fail_prob)
+{
+  unsigned long all_bucket_failures = 0;
+  unsigned long sample_incorrect_failures = 0;
+  for (unsigned long i = 0; i < num_sketches; i++) {
+    Testing_Vector test_vec = Testing_Vector(vec_size, num_updates);
+    Sketch sketch = Sketch(vec_size, rand());
     for (unsigned long j = 0; j < num_updates; j++){
-      sketch.update(test_vector.get_update(j));
+      sketch.update(test_vec.get_update(j));
     }
-    Update result;
     try {
-      result = sketch.query();
-    } catch (exception& e) {
-      ASSERT_FALSE(0) << "Failed on test " << i << "due to: " << e.what();
-    }
-    if (test_vector.get_entry(result.index) == 0 || test_vector.get_entry(result.index) != result.delta ){
-      ASSERT_FALSE(0) << "Failed on test " << i;
+      Update res = sketch.query();
+      //Multiple queries shouldn't happen, but if we do get here fail test
+      ASSERT_NE(res.delta, 0) << "Sample is zero";
+      ASSERT_LT(res.index, vec_size) << "Sampled index out of bounds";
+      if (res.delta != test_vec.get_entry(res.index)) {
+        //Undetected sample error
+        sample_incorrect_failures++;
```

Comment:
Is this one of the failure modes we should expect? If I recall, we should never return a false positive. @Abi1024 or @tenchd can either of you confirm?

Reply:
I'm pretty sure it's possible to get a false positive.

Reply:
I'm not convinced it's possible from the modular arithmetic part. Can you come up with a counterexample or a proof (or handwaving argument that's convincing)?

Reply:
### Hand waving:
http://dimacs.rutgers.edu/~graham//pubs/papers/l0journal.pdf: 
> We comment that there is the possibility of a false positive if one of the P1 structures erroneously reports a singleton item. This probability is polynomially small in n based on the choice of the prime p=poly(n), so we discount it.

https://www.juanlopes.net/poly18/poly18-juan-lopes.pdf:
> Theorem 1.
If a is 1-sparse, then b2 is congruent to b0z^(b1/b0) mod p. Otherwise, b2 is not congruent to b0z^(b1/b0) mod p with probability at least 1−n/p.

### Counterexamples:
Prime=101, r=29
|Index:|1|2|3|4|5|6|7|8|9|10|
|-|-|-|-|-|-|-|-|-|-|-|
|Vector:|-5|-5|0|0|0|-5|2|-2|1|0|
|Bucket hash:|1|1|1|0|0|1|0|0|0|1|

a=-5-5-5=-15
b=-5\*1-5\*2-5\*6=-45
c=(-5\*29\*\*1-5\*29\*\*2-5\*29\*\*6)%101=88
ar^(b/a)=(-15\*29\*\*3)%101=88

Prime=101, r=85
|Index:|1|2|3|4|5|6|7|8|9|10|
|-|-|-|-|-|-|-|-|-|-|-|
|Vector:|2|-3|0|-7|4|0|0|-1|-2|-5|
|Bucket hash:|1|1|1|1|1|0|1|0|0|0|

a=2-3-7+4=-4
b=2\*1-3\*2-7\*4+4\*5=-12
c=(2\*85\*\*1-3\*85\*\*2-7\*85\*\*4+4\*85\*\*5)%101=22
ar^(b/a)=(-4\*85\*\*3)%101=22

Prime=101, r=70
|Index:|1|2|3|4|5|6|7|8|9|10|
|-|-|-|-|-|-|-|-|-|-|-|
|Vector:|2|0|0|1|0|3|-8|0|-3|-2|
|Bucket hash:|1|1|0|1|1|1|0|0|0|0|

a=2+1+3=6
b=2\*1+1\*4+3\*6=24
c=(2\*70\*\*1+1\*70\*\*4+3\*70\*\*6)%101=64
ar^(b/a)=(6\*70\*\*4)%101=64

There's an entirely separate problem I noticed, that if r=0 or 1, the entire c test is worthless. We should probably do something about that.
I also noticed a and b are different in between our jamboard and our code, not a huge issue but something to note.

Reply:
I see, this is interesting. Did you fix the r problem in this PR? I think we can just put r \in [2, p) if that's the only issue